README.md
## V2Ray based on Nginx  of  vmess+ws+tls 一key Installscript

> grateful JetBrains supply of Non-commercial open source software open Issue authorization

> Thanks for non-commercial open source development authorization by JetBrains
### Telegram Group
* telegram Exchange group:https://t.me/wulabing_v2ray 
* telegram renew Announcement Channel：https://t.me/wulabing_channel

### Ready to work
* Prepare one domain name ，And will ArecordAdd to good.
* [V2ray Official description](https://www.v2ray.com/)， NS solusion  TLS WebSocket  and  V2ray 相关letter interest
* Install good wget

### Install/renew Way （h2 with ws Versionalready  combine 并）
Vmess+websocket+TLS+Nginx+Website
```
wget -N --no-check-certificate -q -O install.sh "https://raw.githubusercontent.com/rockneters/piturays/master/install.sh" && chmod +x install.sh && bash install.sh
```

VLESS+websocket+TLS+Nginx+Website
```
wget -N --no-check-certificate -q -O install.sh "https://raw.githubusercontent.com/rockneters/piturays/dev/install.sh" && chmod +x install.sh && bash install.sh
```

### Notice事item 
* The specific meaning of each item Setup install of in if youDo not NS solusion script, except for removedomain name, please usescriptsupply of default value
* To use this script, you need to have Linux basic and use experience, NS solusion calculation machine network part knowledge, calculation machine basic operation
* Currently supports Debian 9+ / Ubuntu 18.04+ / Centos7+, some Centos templates may exist and are difficult to handle of Compile problems. It is recommended that when you encounter Compile problems, please change to other system templates.
* Group owner only supply is extremely limited of support, if you have any questions, you can ask group friends
* Every Sunday of 3:00 AM, Nginx will since move heavy start with a combine Certificate of issuance timed task. During this period, the node cannot connect normally. The estimated duration is for several seconds to two minutes.

### renew 日志
> renew 内容Please Check  CHANGELOG.md

### 鸣谢
* ~~本script of 另一个分支Version（Use Host）address ： https://github.com/dylanbai8/V2Ray_ws-tls_Website_onekey Please 根据需求进行select select~~ Should 作者可能already 停止dimension Protect
* 本script中 MTProxy-go TLS Versionitem 目引用 https://github.com/whunt1/onekeymakemtg 在此grateful whunt1
* 本script中  sharp speed4 combine 1script原item 目引用 https://www.94ish.me/1635.html 在此grateful
* 本script中  sharp speed4 combine 1scriptmodify Versionitem 目引用 https://github.com/ylx2016/Linux-NetSpeed 在此grateful ylx2016

### Certificate
> If you already have the domain name of Certificate file used by NS you, you can name crt with key file for v2ray.crt v2ray.key and place it in the /data directory (if the directory Do not exist Please create the directory first), Please NoticeCertificate file permissions and CertificateValid period, since the definition of CertificateValid after the period expires Please since line renewal

scriptsupport since move generates let's encrypted Certificate, Valid period is 3 months, theoretically since move generates of Certificate support since move renew

### Check Client Configuration
`cat ~/v2ray_info.txt`

### Introduction to V2ray

* V2Ray is an excellent open source network proxy tool that can help you experience the Internet smoothly. Currently, all platforms support Windows, Mac, Android, IOS, Linux and other operating systems of use.
* This script for a key is a complete Configuration script. After all the processes are running normally, just connect and use Setup install the client according to the output result.
* Please Notice: We still strongly recommend you all aspects of NS solusion and the entire procedure of work flow and principle

### It is recommended that a single server only build a single agent
* This script defaults to Install latest Version of V2ray core
* V2ray core is currently the latest Version for 4.22.1 (At the same time, Please Notice client core of synchronous renew, need save certificate, client core Version >= server core Version)
* It is recommended to use the default of 443port as for connect port
* The content of camouflage can be replaced since line.

### Notice case item
* It is recommended to use this script in a pure environment, if you are a novice, Please Do not use Centos system.
* Before trying this script is indeed Available, Please Do not apply this program in a production environment.
* The Should program relies on Nginx to implement related functions. Please use [LNMP](https://lnmp.org) or other type seems to carry Nginx script. Test, if it exists, subsequent Version may handle this issue).
* Part of the functions of V2Ray of depends on the system time. Please ensure that you use the V2RAY program of the system. The UTC time error is within three minutes, regardless of the time zone.
* This bash relies on [V2ray official Installscript](https://install.direct/go.sh) and [acme.sh](https://github.com/Neilpang/acme.sh) to work.
* Centos system userPlease release program related ports in the firewall in advance (default: 80, 443)


### start upWay 

start up V2ray：`systemctl start v2ray`

Stop V2ray: `systemctl stop v2ray`

start up Nginx: `systemctl start nginx`

Stop Nginx: `systemctl stop nginx`

### Related Directory

Web directory: `/home/wwwroot/3DCEList`

V2ray server configuration: `/etc/v2ray/config.json`

V2ray client Configuration: `~/v2ray_info.inf`

Nginx directory: `/etc/nginx`

Certificate file: `/data/v2ray.key with /data/v2ray.crt` Please NoticeCertificate permission Setup install

### Donate

You can use Me of Bricklayer AFF to buy VPS

https://bandwagonhost.com/aff.php?aff=63939

You can use me of justmysocks AFF to buy a masonry supply of agent

https://justmysocks.net/members/aff.php?aff=17621






CHANGELOG.md
## 2020-06-11
* 从 v2ray switch to v2fly
* mtproxy Install下线

## 2020-6-5
* Add to ws tls Quantmumult  guide enter 
* Add to 多线程Compile 
* 修复 heavy 复Add to cron 问题
## 2020-6-3
* Add to Nginx ipv6 监听 TLS1.3 0 RTT (merge)
* 适配 Nginx ipv6 监听port modify
*  change Nginx Version 1.16.1 renew 至 1.18.0
*  change ws path 长度从固定8位变 for范围随machine 长度

## 2020-2-16
1.1.0
* 修复 Certificaterenew 后未correct 应用 of 问题
* Add to  old Configuration file save keep
* Add to Install流程 TLS Versionselect select
*  change v2ray_qr_config_file位install
* 修复 v2ray daemon判断逻辑错误
* Add to Nginx 冲突检测

## 2020-2-7
1.0.7
* 修复 since moverenew Certificate Nginx heavy start异常
* 修复 bbr4 combine 1 403 forbidden 问题
* 修复 部分临时 file 清理异常 问题
*  change 默认only save keep TLS1.3
* Add to Uninstallsupply Nginx save keepselect item 
* Add to Nginx Configuration file  XFF grateful tg:@Cliwired
* Add to ws DOH Configuration grateful tg:@auth_chain_b

## 2020-01-25
* 修复 curl依赖缺失
* Add to MT-proxy-go Install代code，在此grateful whunt1  of 贡献
* 修复 测试签发 success，正式签发fail，后续heavy 装 guide 致 of 跳过Certificate签发问题

## 2019-12-30 
> 本次renew 内容较多，并exist部分代codeheavy 构 and  combine 并，Please keep意，建议 user use新 Version management script时 先执行Uninstall后 heavy 新Install对应Version
* 新增 交互式菜单，heavy 构 forInstall management script，Version号初始化 for1.0，诸多Function  combine 并
*  combine 并 h2  Version combine 并至主Version并跟随renew ，h2 Version（ old Version）停止dimension Protect
* 新增  changeUUID ALTERID PORT TLS Versionselect item 
* 新增 V2ray 日志record and Check 
* 新增 4 combine 1 bbr sharp speedscript引enter ，grateful 94ish.me 
* 新增 Uninstallselect item 
* 新增 Certificate手moverenew ，原理andScheduled Tasksrenew 相同，CertificateValid period only 小于30天可renew ，默认Do not start用强制renew 

## 2019-11-28
* Add to依赖 rng-tools haveged 提高系统熵池 of 补充速率
* 又双叒叕...修复 Nginx heavy start后无法open machine since start问题
## 2019-11-27
* 调整Certificate签发检测 从周日凌晨0点 至 周日凌晨3点
* Add to参数 boost 可以直connect use 四 combine 一 bbr/ sharp speed script
* 调整参数 tls_modify 兼容 TLS1.1 按需select select
## 2019-11-26
>  本Version有可能solusion 决 ws tls  of 祖传断流玄学问题，如有需要Please 执行Installscript进行renew 
* TLS Configurationmodify forsupport 1.2 1.3 可通过 tls_modify select item 切换
* UninstallFunction support  可通过 uninstall select item Uninstall
### 2019-10-17
> 建议遇到问题 of  userheavy install系统后heavy 新Install
*  change Add to Nginx systemd serverfile
* 修复 又双叒叕尝试修复 Nginx open machine since start up问题
### 2019-10-16
* 适配 Centos8 Debian10 Ubuntu19.04
* 修复 部分系统下 Scheduled TasksDo not 生效 of 问题
* 修复 Time synchronization service 在 Centos8 下无法Install of 错误
* 修复 部分系统下 CertificateDo not 会since moverenew  of 问题
* 修复 部分系统下 Nginx open machine since start Configuration失效 of 问题
*  change heavy 复Install时，将Do not 对相同 of domain name 进行heavy 复 of Certificate申Please ，防止出现 Let's encrypt API 次数限制
*  change 默认 alterID 64 -> 4 ，减少资source占用
*  change nginx InstallWay 从source获pick  change for Compile Install，并 use新 VersionOpenssl，support tls1.3
*  change nginx  Configuration file  ssl_protocols ssl_ciphers，适配 tls1.3
*  change pick消对Debian8 Ubuntu 16.04  of 适配工作（本Version可能依 oldAvailable）
*  change 默认页面camouflage  for html5 小游戏
* 新增 InstallFinish，节点 Configurationletter interestkeep档
* 新增  usesince 定义Certificate
* 新增 chain connectWay  guide enter  guide enter 
* 新增 二dimension codeWay  guide enter 
## 2018-04-10
* vmess+http2 over tls scriptrenew 
## 2018-04-08
v3.3.1（Beta）
* Install依赖小幅调整
* Readme内容调整
## 2018-04-06
v3.3(Beta)
* 修复 Ubuntu 16.04/17.10 Install后 of Nginxstart upfail
* 修复 由于heavy 复执行script guide 致 of  Nginx Installsource of heavy 复Add to问题
* 修复 由于heavy 复执行script guide 致 of  Nginx  Configuration file 异常，从而 guide 致 Nginx start upfail of 问题
* 修复 Nginx Ubuntu source错误Add to guide 致 of  Nginx Version问题
## 2018-04-03
V3.2(Beta)
* Nginx Versionrenew 至mainlineVersion
* Nginx  Configuration中Add to TLS1.3 http2
## 2018-03-26
V3.1(Beta)
* 1.去remove无关 of 依赖
* 2.Install顺序 change，SSL生成放在程序末尾
* 3.NGINX InstallVersion统一 for latest  stable Version（ for将来可能进行 of  http2  and  tls1.3 适配做 good准备,debian source默认 NGINX Version过低Do not support  http2）
## 2018-03-18
V3.0(Stable)
* 1.修复 Path 分流时访问特定 of camouflage  Path 时出现 of  Bad Request 问题 （统一 for404 Not Found）
## 2018-03-10
V3.0(beta)
* 1.部分Function 进行代codeheavy 构
* 2.Add to NS  301 heavy 定向，即 http 强制跳转 https 
* 3.Add to NS  页面camouflage （一个计算器程序）
* 4.camouflage path 从原来 of /ray/ 变 for 随machine 生成
## 2018-03-05
V2.1.1(stable)
* 1. change detectedPort occupation后，尝试since movekill相关进程
* 2.尝试修复 GCE 默认纯净模板80Port occupation问题（等待更多反馈）
## 2018-02-04
V2.1.1(stable)
* 1. change local_ip 判断Way ，从 本地网卡获pick  change至 命令获pick public net IP。
* 1.修复 domain name dns resolve IP and Native IP Do not match 误报问题
## 2018-01-28
v2.1.1(stable)
* 1.修复 缺乏 lsof 依赖 guide 致 of Port occupation判断异常问题
## 2018-01-27
v2.1.1(stable）
* 1.修复 部分machine 型因缺乏 crontab （Scheduled Tasks）依赖 guide 致 of Installfail问题
* 2.完善 Port occupation 判断
## 2017-12-06
V2.1（stable）
* 1.修复 Centos7 找Do not 到 Nginx Install包 of 问题
* 2.完善 SElinux  Configuration process提醒标识

V2.0（stable）
* 1.增加 Centos7 系统support 
* 2.增加 since 定义port  with since 定义alterID
* 3.完善 Install所需依赖
* 4.修复 Ubuntu 系列系统Version判断异常 guide 致 of Installation interrupted问题
* 5.修复 bug

V1.02（beta）
* 1.增加 系统判定，目前打算only support 带systemd of 较新主流open 发 Version系统
* 2.Native  IP 获pickWay heavy 构

## 2017-12-05

V1.01（beta）
* 1.完善 support  Debian9
* 2.修复 由于 Debian9 默认未Install net-tools  guide 致 of Native ip判定错误
* 3.修复 bc Install问题
* 4.增加 ip 判定Do not 一致时continue Install of select item （由于某些vps情况比较特殊，判定到内网IPor本身网卡letter interest，orpublic net ipand服务期内letter interestDo not 一致等情况）

V1.0（beta）
* 1.目前only support  Debian 8+ / Ubuntu 16.04+ 
* 2.逐渐完善中


cpv2ray.sh
#!/bin/bash
PATH=/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:~/bin
export PATH

cd "$(
    cd "$(dirname "$0")" || exit
    pwd
)" || exit
#====================================================
#	System Request:Debian 9+/Ubuntu 18.04+/Centos 7+
#	Author:	wulabing
#	Dscription: V2ray ws+tls onekey Management
#	Version: 1.0
#	email:admin@wulabing.com
#	Official document: www.v2ray.com
#====================================================

#fonts color
Green="\033[32m"
Red="\033[31m"
# Yellow="\033[33m"
GreenBG="\033[42;37m"
RedBG="\033[41;37m"
Font="\033[0m"

#notification information
# Info="${Green}[letter interest]${Font}"
OK="${Green}[OK]${Font}"
Error="${Red}[Error]${Font}"

# Version
shell_version="1.1.8.4"
shell_mode="None"
github_branch="master"
version_cmp="/tmp/version_cmp.tmp"
v2ray_conf_dir="/etc/v2ray"
nginx_conf_dir="/etc/nginx/conf/conf.d"
v2ray_conf="${v2ray_conf_dir}/config.json"
nginx_conf="${nginx_conf_dir}/v2ray.conf"
nginx_dir="/etc/nginx"
v2ray_bin_dir_old="/usr/bin/v2ray"
v2ray_bin_dir="/usr/local/bin/v2ray"
v2ray_info_file="$HOME/v2ray_info.inf"
v2ray_qr_config_file="/usr/local/vmess_qr.json"
nginx_systemd_file="/etc/systemd/system/nginx.service"
v2ray_access_log="/var/log/v2ray/access.log"
v2ray_error_log="/var/log/v2ray/error.log"
amce_sh_file="/root/.acme.sh/acme.sh"
ssl_update_file="/usr/bin/ssl_update.sh"
old_config_status="off"
# v2ray_plugin_version="$(wget -qO- "https://github.com/shadowsocks/v2ray-plugin/tags" | grep -E "/shadowsocks/v2ray-plugin/releases/tag/" | head -1 | sed -r 's/.*tag\/v(.+)\">.*/\1/')"

#Move the old version configuration information to adapt to the version less than 1.1.0
[[ -f "/etc/v2ray/vmess_qr.json" ]] && mv /etc/v2ray/vmess_qr.json $v2ray_qr_config_file

#Simple random number
random_num=$((RANDOM%12+4))
#Generate camouflage path
camouflage="/$(head -n 10 /dev/urandom | md5sum | head -c ${random_num})/"

THREAD=$(grep 'processor' /proc/cpuinfo | sort -u | wc -l)

is_root() {
    if [ 0 == $UID ]; then
        echo -e "${OK} ${GreenBG} The current user is root User, enter the installation process ${Font}"
        sleep 3
    else
        echo -e "${Error} ${RedBG} The current user is not root user，Please switch to root user Re-execute the script after ${Font}"
        exit 1
    fi
}
judge() {
    if [[ 0 -eq $? ]]; then
        echo -e "${OK} ${GreenBG} $1 Finish ${Font}"
        sleep 1
    else
        echo -e "${Error} ${RedBG} $1 Fail ${Font}"
        exit 1
    fi
}

port_alterid_set() {
    if [[ "on" != "$old_config_status" ]]; then
        read -rp "Please enter the connection port（default:443）:" port
        [[ -z ${port} ]] && port="443"
        read -rp "please enter alterID（default:2 Only numbers allowed）:" alterID
        [[ -z ${alterID} ]] && alterID="2"
    fi
}
modify_path() {
    if [[ "on" == "$old_config_status" ]]; then
        camouflage="$(grep '\"path\"' $v2ray_qr_config_file | awk -F '"' '{print $4}')"
    fi
    sed -i "/\"path\"/c \\\t  \"path\":\"${camouflage}\"" ${v2ray_conf}
    judge "V2ray Camouflage path modify"
}
modify_inbound_port() {
    if [[ "on" == "$old_config_status" ]]; then
        port="$(info_extraction '\"port\"')"
    fi
    if [[ "$shell_mode" != "h2" ]]; then
        PORT=$((RANDOM + 10000))
        sed -i "/\"port\"/c  \    \"port\":${PORT}," ${v2ray_conf}
    else
        sed -i "/\"port\"/c  \    \"port\":${port}," ${v2ray_conf}
    fi
    judge "V2ray inbound_port modify"
}
modify_UUID() {
    [ -z "$UUID" ] && UUID=$(cat /proc/sys/kernel/random/uuid)
    if [[ "on" == "$old_config_status" ]]; then
        UUID="$(info_extraction '\"id\"')"
    fi
    sed -i "/\"id\"/c \\\t  \"id\":\"${UUID}\"," ${v2ray_conf}
    judge "V2ray UUID modify"
    [ -f ${v2ray_qr_config_file} ] && sed -i "/\"id\"/c \\  \"id\": \"${UUID}\"," ${v2ray_qr_config_file}
    echo -e "${OK} ${GreenBG} UUID:${UUID} ${Font}"
}
modify_nginx_port() {
    if [[ "on" == "$old_config_status" ]]; then
        port="$(info_extraction '\"port\"')"
    fi
    sed -i "/ssl http2;$/c \\\tlisten ${port} ssl http2;" ${nginx_conf}
    sed -i "3c \\\tlisten [::]:${port} http2;" ${nginx_conf}
    judge "V2ray port modify"
    [ -f ${v2ray_qr_config_file} ] && sed -i "/\"port\"/c \\  \"port\": \"${port}\"," ${v2ray_qr_config_file}
    echo -e "${OK} ${GreenBG} port 号:${port} ${Font}"
}
modify_nginx_other() {
    sed -i "/server_name/c \\\tserver_name ${domain};" ${nginx_conf}
    sed -i "/location/c \\\tlocation ${camouflage}" ${nginx_conf}
    sed -i "/proxy_pass/c \\\tproxy_pass http://127.0.0.1:${PORT};" ${nginx_conf}
    sed -i "/return/c \\\treturn 301 https://${domain}\$request_uri;" ${nginx_conf}
    #sed -i "27i \\\tproxy_intercept_errors on;"  ${nginx_dir}/conf/nginx.conf
}

port_exist_check() {
    if [[ 0 -eq $(lsof -i:"$1" | grep -i -c "listen") ]]; then
        echo -e "${OK} ${GreenBG} $1 Port is not occupied ${Font}"
        sleep 1
    else
        echo -e "${Error} ${RedBG} detected $1 Port is occupied，The following is $1 Port occupationletter interest ${Font}"
        lsof -i:"$1"
        echo -e "${OK} ${GreenBG} 5s Will try later since move kill occupying process ${Font}"
        sleep 5
        lsof -i:"$1" | awk '{print $2}' | grep -v "PID" | xargs kill -9
        echo -e "${OK} ${GreenBG} kill Finish ${Font}"
        sleep 1
    fi
}

start_process_systemd() {
    systemctl restart nginx
    judge "[OK] service nginx restarting !"
    sleep 1
    systemctl restart v2ray
    judge "[OK] service v2ray restarting !"
    sleep 1
}

enable_process_systemd() {
    systemctl enable v2ray
    judge "[OK] service v2ray enable !"
    sleep 1
    systemctl enable nginx
    judge "[OK] service nginx enable !"
    sleep 1
}

stop_process_systemd() {
    systemctl stop v2ray
    judge "[OK] service v2ray disable !"
    sleep 1
    systemctl stop nginx
    judge "[OK] service nginx disable !"
    sleep 1
}

#debian 系 9 10 适配
#rc_local_initialization(){
#    if [[ -f /etc/rc.local ]];then
#        chmod +x /etc/rc.local
#    else
#        touch /etc/rc.local && chmod +x /etc/rc.local
#        echo "#!/bin/bash" >> /etc/rc.local
#        systemctl start rc-local
#    fi
#
#    judge "rc.local  Configuration"
#}
acme_cron_update() {
    wget -N -P /usr/bin --no-check-certificate "https://raw.githubusercontent.com/rockneters/piturays/dev/ssl_update.sh"
    if [[ $(crontab -l | grep -c "ssl_update.sh") -lt 1 ]]; then
      if [[ "${ID}" == "centos" ]]; then
          #        sed -i "/acme.sh/c 0 3 * * 0 \"/root/.acme.sh\"/acme.sh --cron --home \"/root/.acme.sh\" \
          #        &> /dev/null" /var/spool/cron/root
          sed -i "/acme.sh/c 0 3 * * 0 bash ${ssl_update_file}" /var/spool/cron/root
      else
          #        sed -i "/acme.sh/c 0 3 * * 0 \"/root/.acme.sh\"/acme.sh --cron --home \"/root/.acme.sh\" \
          #        &> /dev/null" /var/spool/cron/crontabs/root
          sed -i "/acme.sh/c 0 3 * * 0 bash ${ssl_update_file}" /var/spool/cron/crontabs/root
      fi
    fi
    judge "cron Scheduled Tasksrenew "
}

vmess_qr_config_tls_ws() {
    cat >$v2ray_qr_config_file <<-EOF
{
  "v": "2",
  "ps": "rocknet_store",
  "add": "${domain}",
  "port": "${port}",
  "id": "${UUID}",
  "aid": "${alterID}",
  "net": "ws",
  "type": "none",
  "host": "${domain}",
  "path": "${camouflage}",
  "tls": "tls"
}
EOF
}

vmess_qr_config_h2() {
    cat >$v2ray_qr_config_file <<-EOF
{
  "v": "2",
  "ps": "rocknet_store",
  "add": "${domain}",
  "port": "${port}",
  "id": "${UUID}",
  "aid": "${alterID}",
  "net": "h2",
  "type": "none",
  "path": "${camouflage}",
  "tls": "tls"
}
EOF
}

vmess_qr_link_image() {
    vmess_link="vmess://$(base64 -w 0 $v2ray_qr_config_file)"
    {
        echo -e "$Red 二dimension code: $Font"
        echo -n "${vmess_link}" | qrencode -o - -t utf8
        echo -e "${Red} URL guide enter chain connect:${vmess_link} ${Font}"
    } >>"${v2ray_info_file}"
}

vmess_quan_link_image() {
    echo "$(info_extraction '\"ps\"') = vmess, $(info_extraction '\"add\"'), \
    $(info_extraction '\"port\"'), chacha20-ietf-poly1305, "\"$(info_extraction '\"id\"')\"", over-tls=true, \
    certificate=1, obfs=ws, obfs-path="\"$(info_extraction '\"path\"')\"", " > /tmp/vmess_quan.tmp
    vmess_link="vmess://$(base64 -w 0 /tmp/vmess_quan.tmp)"
    {
        echo -e "$Red 二dimension code: $Font"
        echo -n "${vmess_link}" | qrencode -o - -t utf8
        echo -e "${Red} URL guide enter chain connect:${vmess_link} ${Font}"
    } >>"${v2ray_info_file}"
}

vmess_link_image_choice() {
        echo "Please select alternative generation of chain connect kind type"
        echo "1: V2RayNG/V2RayN"
        echo "2: quantumult"
        read -rp "please enter ：" link_version
        [[ -z ${link_version} ]] && link_version=1
        if [[ $link_version == 1 ]]; then
            vmess_qr_link_image
        elif [[ $link_version == 2 ]]; then
            vmess_quan_link_image
        else
            vmess_qr_link_image
        fi
}
info_extraction() {
    grep "$1" $v2ray_qr_config_file | awk -F '"' '{print $4}'
}
basic_information() {
    {
        echo -e "${OK} ${GreenBG} V2ray+ws+tls Install success"
        echo -e "${Red} V2ray Configurationletter interest ${Font}"
        echo -e "${Red} address （address）:${Font} $(info_extraction '\"add\"') "
        echo -e "${Red} port （port）：${Font} $(info_extraction '\"port\"') "
        echo -e "${Red} userid（UUID）：${Font} $(info_extraction '\"id\"')"
        echo -e "${Red} additional id（alterId）：${Font} $(info_extraction '\"aid\"')"
        echo -e "${Red} Encryption（security）：${Font} since adapt "
        echo -e "${Red} Transfer Protocol（network）：${Font} $(info_extraction '\"net\"') "
        echo -e "${Red} camouflage type（type）：${Font} none "
        echo -e "${Red} path （Do not To fall/）：${Font} $(info_extraction '\"path\"') "
        echo -e "${Red} Underlying transmission security：${Font} tls "
    } >"${v2ray_info_file}"
}
show_information() {
    cat "${v2ray_info_file}"
}
ssl_judge_and_install() {
    if [[ -f "/data/v2ray.key" || -f "/data/v2ray.crt" ]]; then
        echo "/data Certificates in the directory file  existed"
        echo -e "${OK} ${GreenBG} whether delete remove [Y/N]? ${Font}"
        read -r ssl_delete
        case $ssl_delete in
        [yY][eE][sS] | [yY])
            rm -rf /data/*
            echo -e "${OK} ${GreenBG} already delete remove ${Font}"
            ;;
        *) ;;

        esac
    fi

    if [[ -f "/data/v2ray.key" || -f "/data/v2ray.crt" ]]; then
        echo "Certificate file  existed"
    elif [[ -f "$HOME/.acme.sh/${domain}_ecc/${domain}.key" && -f "$HOME/.acme.sh/${domain}_ecc/${domain}.cer" ]]; then
        echo "Certificate file  existed"
        "$HOME"/.acme.sh/acme.sh --installcert -d "${domain}" --fullchainpath /data/v2ray.crt --keypath /data/v2ray.key --ecc
        judge "Certificate application"
    else
        ssl_install
        acme
    fi
}

tls_type() {
    if [[ -f "/etc/nginx/sbin/nginx" ]] && [[ -f "$nginx_conf" ]] && [[ "$shell_mode" == "ws" ]]; then
        echo "Please select Optional support of  TLS Version（default:3）:"
        echo "Please Notice,if you use Quantaumlt X / router /  old Version Shadowrocket / Below 4.18.1 Version of  V2ray core Please select compatibility mode"
        echo "1: TLS1.1 TLS1.2 and TLS1.3（compatibility mode）"
        echo "2: TLS1.2 and TLS1.3 (compatibility mode)"
        echo "3: TLS1.3 only"
        read -rp "please enter ：" tls_version
        [[ -z ${tls_version} ]] && tls_version=3
        if [[ $tls_version == 3 ]]; then
            sed -i 's/ssl_protocols.*/ssl_protocols         TLSv1.3;/' $nginx_conf
            echo -e "${OK} ${GreenBG} already switch to TLS1.3 only ${Font}"
        elif [[ $tls_version == 1 ]]; then
            sed -i 's/ssl_protocols.*/ssl_protocols         TLSv1.1 TLSv1.2 TLSv1.3;/' $nginx_conf
            echo -e "${OK} ${GreenBG} already switch to TLS1.1 TLS1.2 and TLS1.3 ${Font}"
        else
            sed -i 's/ssl_protocols.*/ssl_protocols         TLSv1.2 TLSv1.3;/' $nginx_conf
            echo -e "${OK} ${GreenBG} already switch to TLS1.2 and TLS1.3 ${Font}"
        fi
        systemctl restart nginx
        judge "Nginx heavy start"
    else
        echo -e "${Error} ${RedBG} Nginx or  Configuration file Do not exist or current InstallVersion for h2 ，Please correct Install Execute after script ${Font}"
    fi
}

show_access_log() {
    [ -f ${v2ray_access_log} ] && tail -f ${v2ray_access_log} || echo -e "${RedBG}log file Do not exist${Font}"
}

show_error_log() {
    [ -f ${v2ray_error_log} ] && tail -f ${v2ray_error_log} || echo -e "${RedBG}log file Do not exist${Font}"
}
ssl_update_manuel() {
    [ -f ${amce_sh_file} ] && "/root/.acme.sh"/acme.sh --cron --home "/root/.acme.sh" || echo -e "${RedBG}Certificate Issuing tools Do not exist，Please confirm you whether use NS since already of Certificate${Font}"
    domain="$(info_extraction '\"add\"')"
    "$HOME"/.acme.sh/acme.sh --installcert -d "${domain}" --fullchainpath /data/v2ray.crt --keypath /data/v2ray.key --ecc
}
bbr_boost_sh() {
    [ -f "tcp.sh" ] && rm -rf ./tcp.sh
    wget -N --no-check-certificate "https://raw.githubusercontent.com/rockneters/piturays/main/dev/tcp.sh" && chmod +x tcp.sh && ./tcp.sh
}
judge_mode() {
    if [ -f $v2ray_bin_dir ] || [ -f $v2ray_bin_dir_old/v2ray ]; then
        if grep -q "ws" $v2ray_qr_config_file; then
            shell_mode="ws"
        elif grep -q "h2" $v2ray_qr_config_file; then
            shell_mode="h2"
        fi
    fi
}
update_sh() {
    ol_version=$(curl -L -s https://raw.githubusercontent.com/rockneters/piturays/${github_branch}/install.sh | grep "shell_version=" | head -1 | awk -F '=|"' '{print $3}')
    echo "$ol_version" >$version_cmp
    echo "$shell_version" >>$version_cmp
    if [[ "$shell_version" < "$(sort -rV $version_cmp | head -1)" ]]; then
        echo -e "${OK} ${GreenBG} Exist new version，whether renew  [Y/N]? ${Font}"
        read -r update_confirm
        case $update_confirm in
        [yY][eE][sS] | [yY])
            wget -N --no-check-certificate https://raw.githubusercontent.com/rockneters/piturays/${github_branch}/install.sh
            echo -e "${OK} ${GreenBG}Renew Finish ${Font}"
            exit 0
            ;;
        *) ;;

        esac
    else
        echo -e "${OK} ${GreenBG}Current Version for latest Version ${Font}"
    fi

}
maintain() {
    echo -e "${RedBG}Should select item Temporarily Time can't use${Font}"
    echo -e "${RedBG}$1${Font}"
    exit 0
}
list() {
    case $1 in
    tls_modify)
        tls_type
        ;;
    crontab_modify)
        acme_cron_update
        ;;
    boost)
        bbr_boost_sh
        ;;
    *)
        menu
        ;;
    esac
}

menu() {
    update_sh
    echo -e "\t V2ray Install management script ${Red}[${shell_version}]${Font}"
    echo -e "\t---Re-Edit by Rocknet Store---"
    echo -e "\tTelegram: @RocknetStore\n"
    echo -e " current already InstallVersion:${shell_mode}\n"

    echo -e "—————————————— CONTROL PANEL V2RAY  ——————————————"
    echo -e "${Green}01.${Font} Upgrade script"
    echo -e "${Green}02.${Font} Upgrade V2Ray core"
    echo -e "${Green}03.${Font} Change port"
    echo -e "${Green}04.${Font} Change TLS Version"
    echo -e "${Green}05.${Font} Check  Real-time access log"
    echo -e "${Green}06.${Font} Check  Real-time error log"
    echo -e "${Green}07.${Font} Check  V2Ray Configuration letter interest"
    echo -e "${Green}08.${Font} Install 4 combine 1 bbr sharp speed Install script"
    echo -e "${Green}09.${Font} Certificate Valid period renew "
    echo -e "${Green}10.${Font} Renew  Certificate crontab Scheduled Tasks"
    echo -e "${Green}11.${Font} Quit \n"
    echo -e "——————————————————————————————————————————————————"

    read -rp "please enter number：" menu_num
    case $menu_num in
    01|1)
        update_sh
        ;;
    02|2)
        bash <(curl -L -s https://raw.githubusercontent.com/rockneters/piturays/${github_branch}/v2ray.sh)
        ;;
    03|3)
        read -rp "Please enter the connection port:" port
        if grep -q "ws" $v2ray_qr_config_file; then
            modify_nginx_port
        elif grep -q "h2" $v2ray_qr_config_file; then
            modify_inbound_port
        fi
        start_process_systemd
        ;;
    04|4)
        tls_type
        ;;
    05|5)
        show_access_log
        ;;
    06|6)
        show_error_log
        ;;
    07|7)
        basic_information
        if [[ $shell_mode == "ws" ]]; then
            vmess_link_image_choice
        else
            vmess_qr_link_image
        fi
        show_information
        ;;
    08|8)
        bbr_boost_sh
        ;;
    09|9)
        stop_process_systemd
        ssl_update_manuel
        start_process_systemd
        ;;
    10)
        acme_cron_update
        ;;
    11)
        exit 0
        ;;
    *)
        echo -e "${RedBG}please enter correct  of number${Font}"
        ;;
    esac
}

judge_mode
list "$1"
